pkg/messaging/v1/replayable_request.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/dapr/dapr/utils"
+	streamutils "github.com/dapr/dapr/utils/streams"
+)
+
+// Contain pools of *bytes.Buffer and []byte objects.
+// Used to reduce the number of allocations in replayableRequest for buffers and relieve pressure on the GC.
+var (
+	bufPool = sync.Pool{New: newBuffer}
+	// Minimum capacity for the slices is 2KB
+	bsPool = utils.NewByteSlicePool(2 << 10)
+)
+
+func newBuffer() any {
+	return new(bytes.Buffer)
+}
+
+// replayableRequest is implemented by InvokeMethodRequest and InvokeMethodResponse
+type replayableRequest struct {
+	data             io.Reader
+	replay           *bytes.Buffer
+	lock             sync.Mutex
+	currentTeeReader *streamutils.TeeReadCloser
+	currentData      []byte
+}
+
+// WithRawData sets message data.
+func (rr *replayableRequest) WithRawData(data io.Reader) {
+	rr.lock.Lock()
+	defer rr.lock.Unlock()
+
+	if rr.replay != nil {
+		// We are panicking here because we can't return errors
+		// This is just to catch issues during development however, and will never happen at runtime
+		panic("WithRawData cannot be invoked after replaying has been enabled")
+	}
+
+	rr.data = data
+}
+
+// SetReplay enables replaying for the data stream.
+func (rr *replayableRequest) SetReplay(enabled bool) {
+	rr.lock.Lock()
+	defer rr.lock.Unlock()
+
+	if !enabled {
+		rr.closeReplay()
+	} else if rr.replay == nil {
+		rr.replay = bufPool.Get().(*bytes.Buffer)
+		rr.replay.Reset()
+	}
+}
+
+// RawData returns the stream body.
```

Comment:
RawData returns the internal stream which is the replayable one (if replayable is enabled). It will send whatever data has been received, first from the replayable buffer, and then from the stream.